File: tree/bst/bst.go
package bst

import (
	"fmt"
	"math"
)

type Node struct {
	value       int
	left, right *Node
}

type Tree struct {
	root *Node
}

func CreateBinarySearchTree(arr []int) *Tree {
	t := new(Tree)
	size := len(arr)
	t.root = createBinarySearchTreeUtil(arr, 0, size-1)
	return t
}

func createBinarySearchTreeUtil(arr []int, start int, end int) *Node {
	if start <= end {
		curr := new(Node)
		mid := (start + end) / 2
		curr.value = arr[mid]
		curr.left = createBinarySearchTreeUtil(arr, start, mid-1)
		curr.right = createBinarySearchTreeUtil(arr, mid+1, end)
		return curr
	} else {
		return nil
	}
}

func (t *Tree) Add(value int) {
	t.root = addNode(t.root, value)
}

func addNode(n *Node, value int) *Node {
	if n == nil {
		return &Node{value, nil, nil}
	} else {
		if value < n.value {
			return addNode(n.left, value)
		} else {
			return addNode(n.right, value)
		}
	}
}

func (t *Tree) PrintInOrder() {
	if t.root != nil {
		printInOrder(t.root)

	} else {
		fmt.Print("Empty Tree")
	}
}

func printInOrder(n *Node) {
	if n == nil {
		return
	}
	printInOrder(n.left)
	fmt.Print(n.value, " ")
	printInOrder(n.right)
}

func (t *Tree) Find(value int) bool {
	return traverseToFind(t.root, value)
}

func traverseToFind(n *Node, value int) bool {
	if n == nil {
		return false
	}
	if n.value == value {
		return true
	}
	return traverseToFind(n.left, value) || traverseToFind(n.right, value)
}

func (t *Tree) FindMinNode() *Node {
	var node *Node = t.root
	for node != nil && node.left != nil {
		node = node.left
	}
	return node
}

func (t *Tree) FindMaxNode() *Node {
	var node *Node = t.root
	for node != nil && node.right != nil {
		node = node.right
	}
	return node
}

func (t *Tree) IsBST() bool {
	return IsBST(t.root, math.MinInt32, math.MaxInt32)
}

func IsBST(curr *Node, min int, max int) bool {
	if curr == nil {
		return true
	}
	if curr.value < min || curr.value > max {
		return false
	}
	return IsBST(curr.left, min, curr.value) && IsBST(curr.right,
		curr.value, max)
}

func (t *Tree) DeleteNode(node *Node, value int) *Node {
	var temp *Node = nil
	if node != nil {
		if node.value == value {
			if node.left == nil && node.right == nil {
				return nil
			}
			if node.left == nil {
				temp = node.right
				return temp
			}
			if node.right == nil {
				temp = node.left
				return temp
			}
			maxNode := t.FindMax(node.left)
			maxValue := maxNode.value
			node.value = maxValue
			node.left = t.DeleteNode(node.left, maxValue)
		} else {
			if node.value > value {
				node.left = t.DeleteNode(node.left, value)
			} else {
				node.right = t.DeleteNode(node.right, value)
			}
		}
	}
	return node
}

File: tree/bst/bst.cc

#include "bst.hh"

bst::bst(int arr[]) {
  int len = sizeof(arr)/sizeof(arr[0]);
  this->root = createBst(arr, 0, len - 1);
}

bst::~bst() {
  this->root = nullptr;
}

node* bst::createBst(int arr[], int start, int end) {
  if (start <= end) {
    int mid = (start + end) / 2;
    root = new(node);
    root->key = arr[mid];
    root->left = createBst(arr, start, mid - 1);
    root->right = createBst(arr, mid + 1, end);
    return root;
  } else {
    return nullptr;
  }
}

node* bst::addNode(node* root, int key) {
  if (root == nullptr) {
    node* tmp = new(node);
    tmp->key = key;
    tmp->left = nullptr;
    tmp->right = nullptr;
    return tmp;
  }
  if (key < root->key) {
    return addNode(root->left, key);
  } else {
    return addNode(root->right, key);
  }
}

void bst::add(int key) {
  this->root = addNode(this->root, key);
}

node* bst::maxValueNode(node* n) {
  node* current = n;
  while (current && current->right != nullptr) {
    current = current->right;
  }
  return current;
}

node* bst::minValueNode(node* n) {
  node* current = n;
  while (current && current->left != nullptr) {
    current = current->left;
  }
  return current;
}

/*
 * possiblities:
 * 1. Key does not exist
 * 2. Node is leaf node (no children)
 * 3. Node has one child (eighter left or right)
 * 4. Node has both childrens.
*/
node* bst::deleteNode(node* root, int key) {
  if (root == nullptr) {
    return root;
  }

  if (key < root->key) {
    root->left = deleteNode(root->left, key);
  } else if (key > root->key) {
    root->right = deleteNode(root->right, key);
  } else {
    if (root->left == nullptr) {
      node* temp = root->right;
      delete root;
      return temp;
    } else if (root->right == nullptr) {
      node* temp = root->left;
      delete root;
      return temp;
    }
    // Node with two children so
    // get the inorder successor
    // (smallest in the right subtree)
    node* temp = minValueNode(root->right);
    root->key = temp->key;
    root->right = deleteNode(root->right, temp->key);
  }
  return root;
}

void bst::remove(int key) {
  this->root = deleteNode(this->root, key);
}

void bst::removeSubTree(node* root) {
  if(root == nullptr) {
    return;
  }
  removeSubTree(root->left);
  removeSubTree(root->right);
  delete(root);
}

node* bst::findNode(node* root, int key) {
  if (root == nullptr) {
    return nullptr;
  }
  if (root->key == key) {
    return root;
  }
  if (key < root->key) {
    return findNode(root->left, key);
  } else {
    return findNode(root->right, key);
  }
}

node* bst::find(int key) {
  return findNode(this->root, key);
}

node* bst::trimTreeUtil(node* root, int min, int max) {
  if (root == nullptr) return nullptr;

  root->left = trimTreeUtil(root->left, min, max);
  root->right = trimTreeUtil(root->right, min, max);

  if (root->key < min) {
    node* rightSub = root->right;
    delete root;
    return rightSub;
  } else if (root->key > max) {
    node* leftSub = root->left;
    delete root;
    return leftSub;
  }
  return root;
}

void bst::trimTree(int min, int max) {
  this->root = trimTreeUtil(this->root, min, max);
}